### PR TITLE
Using a docker container for clang-cl

### PR DESCRIPTION
**Pull request by lemire**
To fix the clang-format version, use a Docker container.

See https://github.com/RoaringBitmap/CRoaring/pull/577#issuecomment-1917025739

### CODE REVIEW

File: tools/run-clangclcheckdocker.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env bash
+
+set -o noglob
+SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
+MAINSOURCE=$SCRIPTPATH/..
+
+ALL_CROARING_FILES="
```
Comment (SLieve) [inline tools/run-clangclcheckdocker.sh:7]: If possible, wouldn't it make more sense to invert this and define the files we _don't_ want to format? That seems easier to keep up to date.

File: tools/remove-clangcldocker.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+tuser=$(echo $USER | tr -dc 'a-z')
```
Comment (SLieve) [inline tools/remove-clangcldocker.sh:2]: Might want to add `set -e` to fail the script if any line fails.

File: tools/run-clangclcheckdocker.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+echo $container_name
+
+command -v docker >/dev/null 2>&1 || { echo >&2 "Please install docker. E.g., go to https://www.docker.com/products/docker-desktop Type 'docker' to diagnose the problem."; exit 1; }
+
+docker info >/dev/null 2>&1 || { echo >&2 "Docker server is not running? type 'docker info'."; exit 1; }
+
+docker image inspect $container_name >/dev/null 2>&1 || ( echo "instantiating the container" ; docker build --no-cache -t $container_name -f $SCRIPTPATH/clangcldockerfile --build-arg USER_NAME="$tuser"  --build-arg USER_ID=$(id -u)  --build-arg GROUP_ID=$(id -g) .  )
+
+if [ -t 0 ]; then DOCKER_ARGS=-it; fi
+docker run --rm $DOCKER_ARGS -h $container_name -v $MAINSOURCE:$MAINSOURCE:Z  -w $MAINSOURCE  $container_name sh --style=file --verbose --dry-run -i $ALL_CROARING_FILES
```
Comment (SLieve) [inline tools/run-clangclcheckdocker.sh:51]: We should be able to add a "dry run" argument to `run-clangcldocker.sh` and call that script here with the dry run argument.

File: tools/run-clangclcheckdocker.sh (hunk 1 above)
Comment (lemire) [reply tools/run-clangclcheckdocker.sh:7]: @SLieve Can you propose a way to do that?

File: tools/run-clangclcheckdocker.sh (hunk 3 above)
Comment (lemire) [reply tools/run-clangclcheckdocker.sh:51]: Yes. I will make this change.

File: tools/remove-clangcldocker.sh (hunk 2 above)
Comment (lemire) [reply tools/remove-clangcldocker.sh:2]: Done.

File: tools/run-clangclcheckdocker.sh (hunk 3 above)
Comment (lemire) [reply tools/run-clangclcheckdocker.sh]: Done.

File: tools/run-clangclcheckdocker.sh (hunk 1 above)
Comment (SLieve) [reply tools/run-clangclcheckdocker.sh]: We'd have to generate the list of files dynamically again. Perhaps something like:
```
find $MAINSOURCE -regextype posix-extended -regex '.*\.(h|hh|c|cc|cpp)' | grep -vFf clang-format-ignore.txt
```

Where `clang-format-ignore.txt` contains something like:
```
microbenchmarks/toni_ronnko_dirent.h
microbenchmarks/performancecounters/apple_arm_events.h
...
```

Comment (SLieve) [reply tools/run-clangclcheckdocker.sh]: Although that would also find files ignored by git. The original approach modified a bit is probably better:
```
git ls-tree --full-tree --name-only -r HEAD | grep -e ".*\.\(c\|h\|cc\|cpp\|hh\)\$" | grep -vFf clang-format-ignore.txt
```

Comment (lemire) [reply tools/run-clangclcheckdocker.sh]: Done. Please re-review.

File: tools/run-clangcldocker.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+set -e
+set -o noglob
+COMMAND=$@
+SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
+MAINSOURCE=$SCRIPTPATH/..
+ALL_CROARING_FILES=$(cd $SCRIPTPATH/.. && git ls-tree --full-tree --name-only -r HEAD | grep -e ".*\.\(c\|h\|cc\|cpp\|hh\)\$" | grep -vFf clang-format-ignore.txt)
```
Comment (Dr-Emann) [inline tools/run-clangcldocker.sh:7]: nit: this `cd` could use `$MAINSOURCE`

File: tools/run-clangcldocker.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+echo $container_name
+
+command -v docker >/dev/null 2>&1 || { echo >&2 "Please install docker. E.g., go to https://www.docker.com/products/docker-desktop Type 'docker' to diagnose the problem."; exit 1; }
+
+docker info >/dev/null 2>&1 || { echo >&2 "Docker server is not running? type 'docker info'."; exit 1; }
+
+docker image inspect $container_name >/dev/null 2>&1 || ( echo "instantiating the container" ; docker build --no-cache -t $container_name -f $SCRIPTPATH/clangcldockerfile --build-arg USER_NAME="$tuser"  --build-arg USER_ID=$(id -u)  --build-arg GROUP_ID=$(id -g) .  )
+
+if [ -t 0 ]; then DOCKER_ARGS=-it; fi
+docker run --rm $DOCKER_ARGS -h $container_name -v $MAINSOURCE:$MAINSOURCE:Z  -w $MAINSOURCE $container_name sh --style=file --verbose -i $ALL_CROARING_FILES $COMMAND
```
Comment (Dr-Emann) [inline tools/run-clangcldocker.sh:20]: `sh` shouldn't be here: `kszonek/clang-format-17` sets the entrypoint as `clang-format`, so everything after the container name is passed to clang-format, so this is calling `clang-format sh ...`

**Review by Dr-Emann — COMMENTED** [review]
I'd like to consider trying to use a local clang-format if available (and the right version)?

Something like

```sh
if clang-format-17 --version | grep -qF 'version 17.'; then
  clang_format=( clang-format-17 )
elif clang-format --version | grep -qF 'version 17.'; then
  clang_format=( clang-format )
else
  if [ -t 0 ]; then DOCKER_ARGS=-it; fi
  clang_format=( docker run --rm $DOCKER_ARGS -v "$MAINSOURCE:/app:Z" -w "/app" -u "$(id -u $USER):$(id -g $USER)" kszonek/clang-format-17 )
fi

"${clang_format[@]}" "$@"
```

Then we can use that script as if it were `clang-format` itself, passing e.g. `tools/clang-format.sh --style=file --verbose -i -- $ALL_CROARING_FILES` in a wrapper

File: tools/run-clangcldocker.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+set -e
+set -o noglob
+COMMAND=$@
+SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
+MAINSOURCE=$SCRIPTPATH/..
+ALL_CROARING_FILES=$(cd $SCRIPTPATH/.. && git ls-tree --full-tree --name-only -r HEAD | grep -e ".*\.\(c\|h\|cc\|cpp\|hh\)\$" | grep -vFf clang-format-ignore.txt)
+tuser=$(echo $USER | tr -dc 'a-z')
+
+container_name=${CONTAINER_NAME:-"clang-format-for-$tuser"}
```
[the remark this answers is not in the document]
Comment (lemire) [reply tools/run-clangcldocker.sh]: See next commit.

File: tools/run-clangcldocker.sh (hunk 5 above)
Comment (lemire) [reply tools/run-clangcldocker.sh]: See next commit.

File: tools/run-clangcldocker.sh (hunk 4 above)
Comment (lemire) [reply tools/run-clangcldocker.sh]: See next commit.

**Comment by lemire** [comment]
@Dr-Emann Please re-review.

File: tools/run-clangcldocker.sh [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+COMMAND=$@
+SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
+MAINSOURCE=$SCRIPTPATH/..
+ALL_CROARING_FILES=$(cd $MAINSOURCE && git ls-tree --full-tree --name-only -r HEAD | grep -e ".*\.\(c\|h\|cc\|cpp\|hh\)\$" | grep -vFf clang-format-ignore.txt)
+
+if clang-format-17 --version | grep -qF 'version 17.'; then
+  clang-format-17 "$@"
+  exit 0
+elif clang-format --version | grep -qF 'version 17.'; then
+  clang_format=( clang-format )
```
Comment (Dr-Emann) [inline tools/run-clangcldocker.sh:12]: This is unused

File: tools/run-clangcldocker.sh [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  clang-format "$@"
+  exit 0
+fi
```
Comment (Dr-Emann) [inline tools/run-clangcldocker.sh:15]: It looks like we don't pass `--style=file --verbose -i $ALL_CROARING_FILES` if we use clang-format locally? I think we should use the same arguments in all the cases

File: tools/run-clangcldocker.sh [hunk 9]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+set -e
+COMMAND=$@
```
Comment (Dr-Emann) [inline tools/run-clangcldocker.sh:3]: nit: This should [technically](https://www.shellcheck.net/wiki/SC2124) be `$*`

File: tools/run-clangcldocker.sh (hunk 7 above)
Comment (lemire) [reply tools/run-clangcldocker.sh:12]: Removed.

File: tools/run-clangcldocker.sh (hunk 8 above)
Comment (lemire) [reply tools/run-clangcldocker.sh:15]: Corrected.

File: tools/run-clangcldocker.sh (hunk 9 above)
Comment (lemire) [reply tools/run-clangcldocker.sh:3]: Modified.

**Comment by lemire** [comment]
@Dr-Emann I have updated the script.

File: tools/run-clangcldocker.sh [hunk 10]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env bash
+set -e
+COMMAND=$*
+SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
+MAINSOURCE=$SCRIPTPATH/..
+ALL_CROARING_FILES=$(cd $MAINSOURCE && git ls-tree --full-tree --name-only -r HEAD | grep -e ".*\.\(c\|h\|cc\|cpp\|hh\)\$" | grep -vFf clang-format-ignore.txt)
+
+if clang-format-17 --version  2>/dev/null | grep -qF 'version 17.'; then
+  cd $MAINSOURCE; clang-format-17 --style=file --verbose -i "$@" $ALL_CROARING_FILES
+  exit 0
```
Comment (Dr-Emann) [inline tools/run-clangcldocker.sh:10]: It would probably be better to `exit $?` so we keep the return value, same below

**Review by Dr-Emann — APPROVED** [review]
Had a comment, but don't want to hold things up for it

File: tools/run-clangcldocker.sh (hunk 10 above)
Comment (lemire) [reply tools/run-clangcldocker.sh:10]: We have

```
set -e
```

So my expectation is that if the command is in error, then we will never make it to `exit 0`. Am I wrong?

Comment (Dr-Emann) [reply tools/run-clangcldocker.sh:10]: Ah, yeah, that should cover it, yeah.

**Comment by lemire** [comment]
Merging.

**Pull request merged**